### PR TITLE
[FSDP2] enable per-param mesh FSDP2 for MoE

### PR DESCRIPTION
**Pull request by weifengpy**
command: `NGPU=8 MODULE=deepseek_v3 CONFIG=deepseek_v3_16b ./run_train.sh --training.steps 20 --parallelism.expert_parallel_degree 4`

fsdp2 support per-param mesh: https://github.com/pytorch/pytorch/pull/173509

this PR applies fully_shard on transformer_block, sharding experts on `edp_mesh`, and other params on `dp_mesh`. rebased on latest main and handles ep size == 1 carefully 

Screenshot 2026-03-26 at 15 29 31

loss and grad_norm are identical using 10 decimal
Screenshot 2026-03-26 at 17 11 17

Each layer issues 2 AGs and 2 RS, back-to-back. AG and RS overlapps if one is on dp mesh and another on efsdp mesh
Screenshot 2026-03-26 at 17 13 42

active memory are exactly the same. reserved memory are tiny tiny higher
Screenshot 2026-03-26 at 17 16 40

### CODE REVIEW

**Comment by weifengpy** [comment]
those CI errors because we need to land https://github.com/pytorch/pytorch/pull/173509 first

**Comment by xmfan** [comment]
@tianyu-l  We're still gonna need the fine-grained workarounds at least for mxfp8, due to https://github.com/pytorch/torchtitan/pull/2250#discussion_r2713296483. Until that issue is fixed (dynamo bwd tracing of autograd functions), we will trace wrong bwd graph.

**Comment by tianyu-l** [comment]
@xmfan oh, I didn't know. I marked it as high priority for now.

**Comment by weifengpy** [comment]
> Also, not sure if it's blocked by the issue, but we should modify apply_compile to remove the previous workaround fine-grained compilation code.

I didn't change apply_compile because of https://github.com/pytorch/torchtitan/pull/2341.   https://github.com/pytorch/torchtitan/pull/2250#discussion_r2713296483 is new to me

File: torchtitan/models/llama4/parallelize.py [hunk 1]
```diff
@@ -374,40 +378,61 @@ def apply_fsdp(
...
-        # - EP already shards the routed experts on dim 0 (num_experts).
-        #   When FSDP degree > num_experts, default dim-0 sharding causes
-        #   inefficiency due to padding, so we shard on dim-1 (hidden_dim) instead.
-        if transformer_block.moe_enabled:
-            if ep_degree > 1:
-                efsdp_config = fsdp_config.copy()
-                efsdp_config["mesh"] = edp_mesh
-                assert edp_mesh is not None
-                efsdp_ep_size = edp_mesh["efsdp"].size() * ep_degree
-            else:
```
Comment (tianyu-l) [inline torchtitan/models/llama4/parallelize.py:390]: NOTE: there's some recent logic change (fix), which applies to when ep_degree=1. Please adapt.

Comment (weifengpy) [reply torchtitan/models/llama4/parallelize.py:390]: good catch! I am adding those logic back

Comment (weifengpy) [reply torchtitan/models/llama4/parallelize.py:390]: updated the PR to respect latest ep_degree=1 behavior

Image

Comment (weifengpy) [reply torchtitan/models/llama4/parallelize.py:390]: I am taking a look at ci errors if it's related

Comment (weifengpy) [reply torchtitan/models/llama4/parallelize.py:390]: CI error is gone after rebasing on top of https://github.com/pytorch/torchtitan/pull/2723

**Comment by weifengpy** [comment]
I am merging now. I assume rocm CI are still down, since another PR also didn't get result: https://github.com/pytorch/torchtitan/pull/2707
